test/units/plugins/action/test_gather_facts.py

```diff
@@ -45,15 +46,66 @@ def setUp(self):
     def tearDown(self):
         pass
 
+    @patch.object(module_common, '_get_collection_metadata', return_value={})
+    def test_network_gather_facts_smart_facts_module(self, mock_collection_metadata):
+        self.fqcn_task_vars = {'ansible_network_os': 'ios'}
+        self.task.action = 'gather_facts'
+        self.task.async_val = False
+        self.task.args = {}
+
+        plugin = ActionModule(self.task, self.connection, self.play_context, loader=None, templar=self.templar, shared_loader_obj=plugin_loader)
+        get_module_args = MagicMock()
+        plugin._get_module_args = get_module_args
+        plugin._execute_module = MagicMock()
+
+        res = plugin.run(task_vars=self.fqcn_task_vars)
+
+        # assert the gather_facts config is 'smart'
+        facts_modules = C.config.get_config_value('FACTS_MODULES', variables=self.fqcn_task_vars)
+        self.assertEqual(facts_modules, ['smart'])
+
+        # assert the correct module was found
+        self.assertEqual(get_module_args.call_count, 1)
+
+        self.assertEqual(
+            get_module_args.call_args.args,
+            ('ansible.legacy.ios_facts', {'ansible_network_os': 'ios'},)
+        )
+
+    @patch.object(module_common, '_get_collection_metadata', return_value={})
+    def test_network_gather_facts_smart_facts_module_fqcn(self, mock_collection_metadata):
+        self.fqcn_task_vars = {'ansible_network_os': 'cisco.ios.ios'}
+        self.task.action = 'gather_facts'
+        self.task.async_val = False
+        self.task.args = {}
+
+        plugin = ActionModule(self.task, self.connection, self.play_context, loader=None, templar=self.templar, shared_loader_obj=plugin_loader)
```

Comment:
I'm curious why it's necessary to pass the loader in this case (and the others in thise PR)? I was under the impression that `shared_loader_obj` was no longer needed.

Reply:
Hm... I didn't know shared_loader_obj was going away. It's not necessary for this particular test since I've mocked the action plugin's _get_module_args method, but for the others it's an integral part of the change (the action plugins used the shared loader to get the redirect list). I just assumed - should I import plugin_loader in the action plugins instead?

Reply:
Meh, what you have is fine for now- they're basically singletons already, so that constructor arg is probably superfluous, but I don't think there are any active plans to kill it... Just IIUC whatever reason it existed for in the first place doesn't exist anymore. So this looks good to merge to me.

Reply:
I'll get rid of it again anyway in the action_groups PR :grinning: